lib/testing/gpuSimulateN2kPL1bitCorr.cpp

```diff
@@ -0,0 +1,212 @@
+    _samples_per_data_set = config.get<int32_t>(unique_name, "samples_per_data_set");
+    _sub_integration_ntime = config.get<int32_t>(unique_name, "sub_integration_ntime");
+    _blocksize = 8;
+
+    input_plmask_buf = get_buffer("in_plmask_buf");
+    input_plmask_buf->register_consumer(unique_name);
+    input_rfimask_buf = get_buffer("in_rfimask_buf");
+    input_rfimask_buf->register_consumer(unique_name);
+    output_buf = get_buffer("out_buf");
+    output_buf->register_producer(unique_name);
+}
+
+gpuSimulateN2kPL1bitCorr::~gpuSimulateN2kPL1bitCorr() {}
+
+void gpuSimulateN2kPL1bitCorr::main_thread() {
+
+    int input_pl_frame_id = 0;
+    int input_rfi_frame_id = 0;
+    int output_frame_id = 0;
+
+    while (!stop_thread) {
+        uint64_t* pl_mask =
+            (uint64_t*)input_plmask_buf->wait_for_full_frame(unique_name, input_pl_frame_id);
+        if (pl_mask == nullptr)
+            break;
+        uint64_t* rfi_mask =
+            (uint64_t*)input_rfimask_buf->wait_for_full_frame(unique_name, input_rfi_frame_id);
+        if (rfi_mask == nullptr)
+            break;
+        int32_t* counts = (int32_t*)output_buf->wait_for_empty_frame(unique_name, output_frame_id);
+        if (counts == nullptr)
+            break;
+
+        INFO("Simulating GPU PL Correlator for {:s}[{:d}] and {:s}[{:d}] putting result in "
+             "{:s}[{:d}]",
+             input_plmask_buf->buffer_name, input_pl_frame_id, input_rfimask_buf->buffer_name,
+             input_rfi_frame_id, output_buf->buffer_name, output_frame_id);
+
+        // PL:  bool [num_times/64][num_freq][num_el/8][64]
+        // RFI: bool [num_times/1024][num_freq][1024]
+        //
+        // PL:  u64 [num_times/64][num_freq][num_el/8]
+        // RFI: u64 [num_times/1024][num_freq][16]
+
+        int n_integrations = _samples_per_data_set / _sub_integration_ntime;
+        int nt64_inner = _sub_integration_ntime / 64;
+        int nf = _num_local_freq;
+        int ne = _num_elements / 8;
+
+        int n_block_lin = ne / _blocksize;
+        int n_blocks = (n_block_lin * (n_block_lin + 1)) / 2;
+
+        int df_pl = ne;
+        int dt_pl = ne * nf;
+        int df_rfi = 16;
+        int dt_rfi = 16 * nf;
+
+        int diin_c = _blocksize;
+        int db_c = _blocksize * _blocksize;
+        int df_c = n_blocks * _blocksize * _blocksize;
+        int dt_c = nf * n_blocks * _blocksize * _blocksize;
+
+        assert(_num_elements % (8 * _blocksize) == 0);
+
+        INFO("Running PL corr with nt_outer={:d}, nt_inner={:d}, num_freq={:d}, "
+             "num_downsampled_elements={:d}",
+             n_integrations, _sub_integration_ntime, nf, ne);
+
+        for (int t_out = 0; t_out < n_integrations; t_out++) {
+            for (int f = 0; f < nf; f++) {
+                int block_idx = 0;
+                for (int i_out = 0; i_out < n_block_lin; i_out++) {
+                    for (int j_out = 0; j_out <= i_out; j_out++) {
```

Comment:
The eagle-eyed among use spot this line. Many others wont. Please add a comment describing what makes this line special.